### PR TITLE
Adding a HAL module debug sink interface.

### PR DESCRIPTION
**Pull request by benvanik**
HAL module creators can now pass in a debug sink that injects callbacks that can be made by the HAL module on request from programs or by the HAL module itself. Today it just has a buffer view trace callback but in the future we could use this to report profiling information or asynchronous errors.

A default `iree_hal_module_debug_sink_stdio(file)` is provided that preserves the current behavior of writing trace buffers to stderr (when HAL string utils are enabled). `iree_hal_module_debug_sink_null()` can be used to ensure no string handling code is compiled in if a caller doesn't want it. Callers can also implement whatever behavior they want.

### CODE REVIEW

File: runtime/src/iree/modules/hal/debugging.c [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// Debug Sink
+//===----------------------------------------------------------------------===//
+
+IREE_API_EXPORT iree_hal_module_debug_sink_t
+iree_hal_module_debug_sink_null(void) {
+  iree_hal_module_debug_sink_t sink = {0};
+  return sink;
+}
+
+static iree_status_t iree_hal_module_buffer_view_trace_stdout(
```
Comment (sogartar) [inline runtime/src/iree/modules/hal/debugging.c]: Shouldn't this be called `iree_hal_module_buffer_view_trace_stderr`? It actually writes to `stderr`.

Comment (benvanik) [reply runtime/src/iree/modules/hal/debugging.c]: good catch!

Comment (benvanik) [reply runtime/src/iree/modules/hal/debugging.c]: I renamed it to stdio and it now takes a file handle so the caller gets to decide :)

File: runtime/src/iree/modules/hal/debugging.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+//===----------------------------------------------------------------------===//
+// Debug Sink
+//===----------------------------------------------------------------------===//
+
+// Receives a
+typedef iree_status_t(
+    IREE_API_PTR* iree_hal_module_buffer_view_trace_callback_fn_t)(
+    void* user_data, iree_string_view_t key, iree_host_size_t buffer_view_count,
+    iree_hal_buffer_view_t** buffer_views, iree_allocator_t host_allocator);
```
Comment (sogartar) [inline runtime/src/iree/modules/hal/debugging.h:26]: Why is `iree_allocator_t host_allocator` an argument, is it really required there? Can't it be a part of `user_data`?

**Review by sogartar — APPROVED** [review]
Looks good.
I would like to plumb this through the Python bindings and start using it.

File: runtime/src/iree/modules/hal/debugging.h (hunk 2 above)
Comment (benvanik) [reply runtime/src/iree/modules/hal/debugging.h:26]: It's the allocator from the HAL module - an implementation could ignore it and use their own in user_data, but this is the only way to get at the HAL module one.

**Pull request merged**